crates/pallet-evm-system/src/migrations/broken_nonces_recovery.rs

```diff
@@ -0,0 +1,189 @@
+                    .map(|account| account.nonce.is_zero())
+                    .unwrap_or(true)
+            })
+            .count();
+        ensure!(
+            accounts_to_recover == 0,
+            "There should be no accounts left for recovery",
+        );
+        Ok(())
+    }
+}
+
+impl<EP, T> MigrationBrokenNoncesRecovery<EP, T>
+where
+    EP: EvmStateProvider<<T as Config>::AccountId>,
+    T: Config<AccountId = H160>,
+    <T as Config>::Index: rlp::Encodable,
+{
+    /// Checks the contract account state and recovers it if necessary.
+    /// - If the state is missing, recreates it.
+    /// - If the state has nonce = 0, writes the smallest possible valid nonce.
+    ///
+    /// Precompiled contracts in Ethereum usually have nonce = 0 as well. However, since precompiled
+    /// contracts are typically implemented by hooking calls to specific addresses and adding dummy
+    /// state (to ensure they are callable like regular contracts), there's no need for a non-zero
+    /// nonce unless they explicitly perform state-changing operations like `CREATE`. Therefore,
+    /// precompiled contracts MUST NOT be passed here.
+    fn recover(
+        id: &<T as Config>::AccountId,
+        account: &mut Option<AccountInfo<<T as Config>::Index, <T as Config>::AccountData>>,
+    ) -> Weight {
+        let Some(account) = account.as_mut() else {
+            info!("Account {id} lacks its state");
+            let (nonce, weight) = Self::min_nonce(id);
+            *account = Some(AccountInfo {
+                nonce,
+                data: Default::default(),
+            });
+            return weight;
+        };
+        if !account.nonce.is_zero() {
+            return Default::default();
+        }
+        info!("Account {id} has zero nonce");
+        let (nonce, weight) = Self::min_nonce(id);
+        account.nonce = nonce;
+        weight
+    }
+
+    /// Computes the minimum possible nonce for a given account.
+    fn min_nonce(id: &<T as Config>::AccountId) -> (<T as Config>::Index, Weight) {
+        let mut weight = Weight::default();
+        let mut nonce = <T as Config>::Index::one();
+        while {
+            let contract_id = contract_address(id, nonce);
+            let (has_code, w) = EP::has_code(&contract_id);
+            weight.saturating_accrue(w);
+            let has_state = <Account<T>>::contains_key(contract_id);
+            weight.saturating_accrue(T::DbWeight::get().reads(1));
+            has_code || has_state
+        } {
+            nonce = nonce
+                .checked_add(&One::one())
+                .expect("Nonce value mustn't overflow");
+        }
+        info!("Account {id} minimal valid nonce is {nonce:?}");
+        (nonce, weight)
+    }
+}
+
+/// Contract address that will be produced by the [`CREATE` opcode][1].
+///
+/// [1]: https://ethereum.github.io/yellowpaper/paper.pdf#section.7
+fn contract_address(sender: &H160, nonce: impl rlp::Encodable) -> H160 {
```

Comment:
In case we use own implementation, what about moving this to `crypto-utils` or `crypto-utils-evm`? It would be helpful to reuse it later.

Reply:
But it would kind of is also be fine to leave here - in a sense, this is a highly specialized conversion for this migration; we wouldn't want it to be updated/changed if the shared implementation changes.

Reply:
I'm for `evm-utils`. I also think it'll be useful after the second use case.